accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+
+**Located in xamarin/xamarin-macios repo**
+```c#
+// ios SDK part
+public class CFNetworkHandler : HttpClientHandler
+{
+  public CFNetworkHandle (): base (true)
+  {
+  }
+
+  public override bool SupportsProxy => objc.SendMessage ("some data") as bool;
+
+  // Every property would have to have override
+}
+```
+
+**Disadvantages**
+
+* Testing will have to be done at the integration level.
+* The existing handlers would need to be changed to derive from common type the proxy could operate on (HttpClientHandler, an interface, or something else).
+* All HttpClientHandler properties will have to be virtual or we’ll have to introduce IHttpClientHandler which would also solve the recursion problem.
+* Every method and property will have to have implementation override otherwise will hang in recursion / stack overflow.
+* Factory for creating the platform instance won’t be part of public contract.
+* Adding new public members will be difficult without IHttpClientHandler.
+
+**Conclusion**
+
+This is the cheapest option to implement which meets our needs as it’s closest to what we Mono and Xamarin use today.
+
+### Platform Handler Construction
+This subsection covers various options we have to construct the underlying platform instance.
+
+**Proposal 2a: ILLinker Magic**
+
+Each SDK would write optional ILLinker custom step which would rewrite construction logic for platform instance with implementation which is controlled by SDK.
+```c#
+// System.Net.Http.dll (for iOS, Android, etc)
+
+public class HttpClientHandler
+{
+  readonly HttpClientHandler platformInstance;
+
+  public HttpClientHandler ()
+  {
+   if (GetType() == type (HttpClientHandler))
+     platformInstance = CreateNewInstance ();
+  }
+
+  // ILinker will recognize this method and change the body to be platform specific
+  private static HttpClientHandler CreateNewInstance ()
+  {
+    throw new PlatformNotSupportedException();    
+  }
+}
+```
+
+**Disadvantages**
+
+* Requires ILLinker to always run which slows down build process.
+* Changing internal state of HttpClientHandler will be tricky as it can break hook used by ILLinker.
+
+**Proposal 2b: Use Reflection**
+
+The construction code would use reflection call to static method for platform where the late binding is required.
+```c#
+// System.Net.Http.dll
+
+public class HttpClientHandler
+{
+  #if iOS
+  [PreserveDependency ("Create", "HttpClientHandlerFactory", "Xamarin.iOS")
+  #endif
+  public HttpClientHandler ()
+  {
+    platformInstance = Type.GetType ("HttpClientHandlerFactory").GetMethod("Create").Invoke () as HttpClientHandler;
+  }
+}
+```
+
+**Disadvantages**
+
+* Costly and slow reflection infrastructure would be pulled in.
+* Need to add PreserveDependency to all possible target assemblies.
```

Comment:
They are only issues if this was the only option.
If you use reflection (and I rather not) for debug/linker-less builds then you save more time (linking) than the execution time (for reflection).

For release/optimized builds you want the linker and that removes the disadvantages.

Reply:
That's what the text says already `Costly and slow reflection infrastructure would be pulled in`